feat(search): Similarity search for Opensearch and Elasticsearch

frascuchon opened the pull request:
<!-- Thanks for your contribution! As part of our Community Growers initiative 🌱, we're donating Justdiggit bunds in your name to reforest sub-Saharan Africa. To claim your Community Growers certificate, please contact David Berenstein in our Slack community or fill in this form https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This PR includes a bunch of changes regarding the `search_engine` module: 

- The base class includes a new method to tackle vector generation and defines the `similarity_search` method.
- Two similar implementations of the `SearchEngine` class have been included `OpenSearchEngine` and `ElasticSearchEngine`. Since they share a lot of functionality, a common abstract class has been implemented for common code blocks, letting each implementation handle its own details.
- A new way for engine registration and generation has been developed. This new way is based on a class decorator. The aim of this approach is to simplify future engine implementations, keeping the code isolated for each new implementation.
The following example shows how a new engine could be implemented and registered as an available engine for the Argilla server:
```python
from argilla.server.search_engine import SearchEngine

@SearchEngine.register(engine_name="my-new-engine")
class NewSearchEngineImplementation(SearchEngine):
   # implemente all required method
   ...
```
Then, you can use your search engine by environment variable:
```bash
ARGILLA_SEARCH_ENGINE=my-new-engine python -m argilla server
```
The `ARGILLA_SEARCH_ENGINE` environment variable has been added as a common server setting. This is a first approach and may change in the future. This is the reason why I did not include it in the docs section.

Refs #3028

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [X] New feature (non-breaking change which adds functionality)
- [X] Refactor (change restructuring the codebase without changing functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

- [ ] Test A
- [ ] Test B

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

frascuchon committed:
feat: Define new method for search engine interface

frascuchon committed:
refactor: Define common base class for elasticsearch and opensearch engines

frascuchon committed:
refactor: Use base class for opensearch engine

frascuchon committed:
feat: Add new elasticsearch engine using the common base class

frascuchon committed:
tests: Update tests for the new search engine system

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
chore: Remove extra method

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

gabrielmbmb commented on src/argilla/server/search_engine/elasticsearch.py:
```diff
@@ -0,0 +1,135 @@
+from argilla.server.settings import settings
+
+
+def _compute_num_candidates_from_k(k: int):
```
```suggestion
def _compute_num_candidates_from_k(k: int) -> int:
```

gabrielmbmb commented on src/argilla/server/search_engine/elasticsearch.py:
```diff
@@ -0,0 +1,135 @@
+    def __post_init__(self):
+        self.client = AsyncElasticsearch(**self.config)
+
+    def _configure_index_settings(self):
```
```suggestion
    def _configure_index_settings(self) -> Dict[str, Any]:
```

gabrielmbmb commented on src/argilla/server/search_engine/elasticsearch.py:
```diff
@@ -0,0 +1,135 @@
+                "index": True,
+                # can similarity property also be part of config @frascuchon ?
+                # relates vector search similarity metric
+                "similarity": "l2_norm",  ## default value regarding the knn best practices es documentation
```
I think that we should let the user configure this when creating settings for a collection of vectors. In some cases, depending on the vectors used, one metric could be better than other.

frascuchon replied:
Yes. We can iterate over it in future PRs. The `similarity` param could affect other default ones.

gabrielmbmb reviewed (commented):
I think that we should allow the user to configure the similarity metric. Also, I think we could assume that no matter the vector database that Argilla is using, HNSW algorithm will be used.

HNSW is the default ANN algorithm in all the vector databases, and I think it's important also to let the user configure the parameters of the algorithm `ef`, `m` and `ef_construction`, as these would affect the results of the search.

frascuchon committed:
Apply suggestions from code review

Co-authored-by: Gabriel Martín Blázquez <gmartinbdev@gmail.com>

gabrielmbmb commented on src/argilla/server/search_engine/base.py:
```diff
@@ -96,3 +96,24 @@ async def search(
+        value: Optional[List[float]] = None,
+        record: Optional[Record] = None,
+        user_response_status_filter: Optional[UserResponseStatusFilter] = None,
+        max_results: conint(ge=2, le=500) = 100,
```
No validation will be applied here by `pydantic`, so let's just use `int`

```suggestion
        max_results: int = 100,
```

frascuchon committed:
feat: Enable dynamic engine registration

frascuchon committed:
refactor: Register current engines using decorator

frascuchon committed:
feat: create engine dynamically depending on the `ARGILLA_SEARCH_ENGINE`  environment variable

frascuchon committed:
tests: Running search engine tests depending on configured engine

frascuchon committed:
ci: Configure search engine for test in ci pipelines

frascuchon committed:
Merge branch 'feauture/similarity-search-for-opensearch-and-elasticsearch' of github.com:argilla-io/argilla into feauture/similarity-search-for-opensearch-and-elasticsearch

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
Apply suggestions from code review

Co-authored-by: Gabriel Martín Blázquez <gmartinbdev@gmail.com>

gabrielmbmb commented on .github/workflows/run-python-tests.yml:
```diff
@@ -77,6 +77,16 @@ jobs:
         env:
           HF_HUB_ACCESS_TOKEN: ${{ secrets.HF_HUB_ACCESS_TOKEN }}
         run: echo "Enable HF access token"
+      - name: Set huggingface hub credentials
```
```suggestion
      - name: Set Argilla search engine env variable
```

gabrielmbmb commented on .github/workflows/run-python-tests.yml:
```diff
@@ -77,6 +77,16 @@ jobs:
+        env:
+          ARGILLA_SEARCH_ENGINE: elasticsearch
+        run: echo "Configure elasticsearch engine"
+      - name: Set huggingface hub credentials
```
```suggestion
      - name: Set Argilla search engine env variable
```

gabrielmbmb commented on src/argilla/server/search_engine/elasticsearch.py:
```diff
@@ -0,0 +1,132 @@
+            config,
+            es_number_of_shards=settings.es_records_index_shards,
+            es_number_of_replicas=settings.es_records_index_replicas,
+        )
```
```suggestion
    @classmethod
    async def new_instance(cls) -> "ElasticSearchEngine":
        config = dict(
            hosts=settings.elasticsearch,
            verify_certs=settings.elasticsearch_ssl_verify,
            ca_certs=settings.elasticsearch_ca_path,
            retry_on_timeout=True,
            max_retries=5,
        )
        return cls(
            config,
            es_number_of_shards=settings.es_records_index_shards,
            es_number_of_replicas=settings.es_records_index_replicas,
        )
```

gabrielmbmb commented on src/argilla/server/search_engine/elasticsearch.py:
```diff
@@ -0,0 +1,132 @@
+    return 2000
+
+
+@SearchEngine.register(engine_name="opensearch")
```
```suggestion
@SearchEngine.register(engine_name="elasticsearch")
```

frascuchon committed:
ci: Fix ci configuration

gabrielmbmb committed:
feat: convert method to `asynccontextmanager` (#3774)

# Description

A small update to rename `SearchEngine.search_engine_by_name` method to
`SearchEngine.get_by_name` to be less redundant and also transform it to
be a context manager, so it could be used outside of the FastAPI context
too.

**Type of change**

- [x] Refactor (change restructuring the codebase without changing
functionality)
- [ ] 
**How Has This Been Tested**

NA

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

frascuchon committed:
Apply suggestions from code review

Co-authored-by: Gabriel Martín Blázquez <gmartinbdev@gmail.com>

frascuchon committed:
ci: Disable running tests on 7.11

frascuchon committed:
ci: Running tests only with fully compatible engines

frascuchon committed:
ci: Setting conditional env vars properly

frascuchon committed:
ci: Disable elasticsearch engine tests
